perun-core/src/main/java/cz/metacentrum/perun/core/impl/UsersManagerImpl.java

```diff
@@ -794,90 +795,112 @@ public void removeAllUserExtSources(PerunSession sess, User user) {
 
 	@Override
 	public List<User> findUsers(PerunSession sess, String searchString) {
-		Set<User> users = new HashSet<>();
-
-		log.debug("Searching for users using searchString '{}'", searchString);
-
-		// Search by mail (member)
-		users.addAll(jdbc.query("select " + userMappingSelectQuery +
-					" from users, members, member_attr_values, attr_names " +
-					"where members.user_id=users.id and members.id=member_attr_values.member_id and member_attr_values.attr_id=attr_names.id and " +
-					"attr_names.attr_name='urn:perun:member:attribute-def:def:mail' and " +
-					"lower(member_attr_values.attr_value)=lower(?)", USER_MAPPER, searchString));
-
-		// Search preferred email (user)
-		users.addAll(jdbc.query("select " + userMappingSelectQuery +
-					" from users, user_attr_values, attr_names " +
-					"where users.id=user_attr_values.user_id and user_attr_values.attr_id=attr_names.id and " +
-					"attr_names.attr_name='urn:perun:user:attribute-def:def:preferredMail' and " +
-					"lower(user_attr_values.attr_value)=lower(?)", USER_MAPPER, searchString));
-
-		// Search logins in userExtSources
-		users.addAll(jdbc.query("select " + userMappingSelectQuery +
-					" from users, user_ext_sources " +
-					"where lower(user_ext_sources.login_ext)=lower(?) and user_ext_sources.user_id=users.id", USER_MAPPER, searchString));
-
-		// Search logins in attributes: login-namespace:*
-		users.addAll(jdbc.query("select distinct " + userMappingSelectQuery +
-					" from attr_names, user_attr_values, users " +
-					"where attr_names.friendly_name like 'login-namespace:%' and lower(user_attr_values.attr_value)=lower(?) " +
-					"and attr_names.id=user_attr_values.attr_id and user_attr_values.user_id=users.id",
-					USER_MAPPER, searchString));
-
-		// Search by userId
-		try {
-			int userId = Integer.parseInt(searchString);
-			users.addAll(jdbc.query("select " + userMappingSelectQuery + " from users where id=?", USER_MAPPER, userId));
-		} catch (NumberFormatException e) {
-			// IGNORE
-		}
-
-		users.addAll(findUsersByName(sess, searchString));
-
-		return new ArrayList<>(users);
+		return findUsers(searchString, false);
 	}
 
 	@Override
 	public List<User> findUsersByExactMatch(PerunSession sess, String searchString) {
-		Set<User> users = new HashSet<>();
+		return findUsers(searchString, true);
+	}
 
-		log.debug("Searching for users using searchString '{}'", searchString);
+	/**
+	 * Returns list of users who matches the searchString, searching name, id, member attributes, user attributes
+	 * and userExtSource attributes (listed in CoreConfig).
+	 *
+	 * @param searchString string used to search by
+	 * @param exactMatch if true, searches name only by exact match
+	 * @return list of users
+	 */
+	private List<User> findUsers(String searchString, boolean exactMatch) {
+		String userNameQueryString;
+		if (exactMatch) {
```

Comment:
Both conditional branches can be extracted to separate methods. One could be called something like "prepareUserSearchQueryExactMatch" and the second one "prepareUserSearchQuerySimilarMatch" or something like that. It will be more readable. Moreover, the second method can be used also in MembersManagerImpl so you eliminate duplicities.

Reply:
Ok, I've created new methods in Utils.